xrootd/mux/mux.go

```diff
@@ -0,0 +1,113 @@
+// Copyright 2018 The go-hep Authors.  All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package mux // import "go-hep.org/x/hep/xrootd/mux"
+
+import (
+	"math"
+	"sync"
+
+	"github.com/pkg/errors"
+	"go-hep.org/x/hep/xrootd/protocol"
+)
+
+type ServerResponse struct {
+	Data []byte
+	Err  error
+}
+
+type dataSendChan chan<- *ServerResponse
+type DataRecvChan <-chan *ServerResponse
+
+const streamIDPartSize = math.MaxUint8
+const streamIDPoolSize = streamIDPartSize * streamIDPartSize
+
+// Mux manages channels by their ids.
+// Basically, it's a map[StreamID] chan<-*ServerResponse
+// with methods to claim, free and pass data to specific channel by id.
+type Mux struct {
+	mu          sync.Mutex
+	dataWaiters map[protocol.StreamID]dataSendChan
+	freeIDs     chan protocol.StreamID
+}
+
+// New creates new Mux
+func New() *Mux {
+	m := Mux{
+		dataWaiters: make(map[protocol.StreamID]dataSendChan),
+		freeIDs:     make(chan protocol.StreamID, streamIDPoolSize),
+	}
+
+	for firstByte := uint16(0); firstByte < streamIDPartSize; firstByte++ {
+		for secondByte := uint16(0); secondByte < streamIDPartSize; secondByte++ {
+			m.freeIDs <- protocol.StreamID{byte(firstByte), byte(secondByte)}
+		}
+	}
+
+	return &m
+}
+
+// Claim searches for unclaimed id and returns corresponding channel
+func (m *Mux) Claim() (id protocol.StreamID, channel DataRecvChan, err error) {
+	bidirectionalChannel := make(chan *ServerResponse)
+
+	for channel == nil && err == nil {
+		select {
+		case id = <-m.freeIDs:
+			m.mu.Lock()
+			if _, claimed := m.dataWaiters[id]; claimed { // Skip id if it was already claimed manually via ClaimWithID
+				m.mu.Unlock()
+				continue
+			}
+
+			m.dataWaiters[id] = bidirectionalChannel
+			m.mu.Unlock()
+			channel = bidirectionalChannel
+		default:
+			err = errors.New("streammanager: cannot Claim free channel")
+		}
+	}
+
+	return
+}
+
+// ClaimWithID checks if id is unclaimed and returns the corresponding channel in case of success
+func (m *Mux) ClaimWithID(id protocol.StreamID) (channel DataRecvChan, err error) {
```

Comment:
ditto, wrt named returned values.

IMHO, this reads as well (if not better :P):
```go
func (m *Mux) ClaimWithID(id protocol.StreamID) (DataRecvChan, error) {
    m.mu.Lock()
    defer m.mu.Unlock()
    ch := make(chan *ServerResponse, 1)

    if _, claimed := m.dataWaiters[id]; claimed {
        return nil, errors.Errorf(...)
    }

    m.dataWaiters[id] = ch
    return ch, nil
}
```